src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/CoffObjectWriter.cs

```diff
@@ -0,0 +1,1097 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Numerics;
+using System.Reflection.PortableExecutable;
+using System.Runtime.InteropServices;
+using System.Text;
+using ILCompiler.DependencyAnalysis;
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+using Internal.TypeSystem.TypesDebugInfo;
+using static ILCompiler.DependencyAnalysis.RelocType;
+using static ILCompiler.ObjectWriter.CoffObjectWriter.CoffRelocationType;
+
+namespace ILCompiler.ObjectWriter
+{
+    /// <summary>
+    /// COFF object file format writer for Windows targets.
+    /// </summary>
+    /// <remarks>
+    /// The PE/COFF object format is described in the official specifciation at
+    /// https://learn.microsoft.com/en-us/windows/win32/debug/pe-format. However,
+    /// numerous extensions are missing in the specification. The most notable
+    /// ones are listed below.
+    ///
+    /// Object files with more than 65279 sections use an extended big object
```

Comment:
What does our test coverage look like for big object vs. regular? I am wondering whether it would be a good idea to use big object unconditionally, even for smaller number of sections.

Reply:
The big object files mainly kick in when the COMDAT sections are used to share method code. While some such sections can be produced in regular single file compilation, it's very limited occurrence and unlikely to hit the 65279 section limit of standard object file. The main trigger is the `--multifile` option (`IlcMultiModule` property in MSBuild), which changes the outcome of [this condition]( https://github.com/dotnet/runtime/blob/fac60d0e972dd3f16ddc568573f2934c5fdc9acc/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ObjectWriter.cs#L960-L967).

The test coverage on Windows is provided by the NativeAot runs with the `BuildNativeAotFrameworkObjects=true` parameter, and specifically by the `MultiModule` smoke test. We essentially get the normal objects tested by the regular smoke tests, and the big ones in the multi-file mode.

The downside of enabling big object files unconditionally is that the the format is not officially documented and many 3rd-party tools don't implement it, or implement it only partially. That makes analyzing the output at object file level cumbersome. Since the `IlcMultiModule` MSBuild property is not set by default and there's no official documentation for it, in vast majority of cases the big object files would not be encountered by the users.